### PR TITLE
TRAD-183 Go SDK update from recent block hash

### PR DESCRIPTION
**Pull request by yobloxroute**

### CODE REVIEW

File: bxserum/provider/grpc.go [hunk 1]
```diff
@@ -63,10 +72,72 @@ func NewGRPCClientWithOpts(opts RPCOpts) (*GRPCClient, error) {
...
-	return &GRPCClient{
-		apiClient:  pb.NewApiClient(conn),
-		privateKey: opts.PrivateKey,
-	}, nil
+	client := &GRPCClient{
+		apiClient:                     pb.NewApiClient(conn),
+		privateKey:                    opts.PrivateKey,
+		recentBlockHashExpiryDuration: opts.CacheBlockHash,
+	}
+	go func() {
```
Comment (aspin) [inline bxserum/provider/grpc.go]: this code also needs to be in `provider/ws.go` –– probably would be a good idea to create a struct encapsulating this logic?

File: bxserum/provider/grpc.go [hunk 2]
```diff
@@ -7,15 +7,24 @@ import (
...
 type GRPCClient struct {
 	pb.UnimplementedApiServer
 
+	mutex *sync.RWMutex
+
 	apiClient  pb.ApiClient
 	privateKey *solana.PrivateKey
+
+	recentBlockHash               *pb.GetRecentBlockHashResponse
+	recentBlockHashTime           *time.Time
```
Comment (aspin) [inline bxserum/provider/grpc.go:26]: any reason this needs to be a pointer? Seems to result in some unnecessary dereferencing

File: bxserum/provider/grpc.go [hunk 3]
```diff
@@ -7,15 +7,24 @@ import (
...
 	"google.golang.org/grpc"
 	"google.golang.org/grpc/credentials/insecure"
+	"sync"
+	"time"
 )
 
 type GRPCClient struct {
 	pb.UnimplementedApiServer
 
+	mutex *sync.RWMutex
```
Comment (aspin) [inline bxserum/provider/grpc.go]: any reason this needs to be a pointer?

File: bxserum/provider/grpc.go (hunk 1 above)
Comment (aspin) [reply bxserum/provider/grpc.go]: It'd also be nice to have a boolean flag in RPCOpts for if the user wants to enable this. Can default to `true`

File: bxserum/provider/grpc.go [hunk 4]
```diff
@@ -63,10 +72,72 @@ func NewGRPCClientWithOpts(opts RPCOpts) (*GRPCClient, error) {
...
+	response := g.getCachedBlockHash()
+	if response != nil {
+		return response, nil
+	}
+
+	g.mutex.Lock()
+	defer g.mutex.Unlock()
+
+	now := time.Now()
+	if g.recentBlockHash != nil && g.recentBlockHashTime.Before(now.Add(g.recentBlockHashExpiryDuration)) {
```
Comment (aspin) [inline bxserum/provider/grpc.go]: think the `.Before` call is backward –– should be `g.recentBlockHashTime.Add(g.recentblockHashExpiryDuration).Before(now)`?

Comment (aspin) [reply bxserum/provider/grpc.go]: `g.recentBlockHashTime` is already always before `now`

**Review by aspin — CHANGES_REQUESTED** [review]

File: bxserum/provider/grpc.go (hunk 2 above)
Comment (yobloxroute) [reply bxserum/provider/grpc.go:26]: just to be sure that we're not returning empty values, but yes, empty time is not valid :)

File: bxserum/provider/grpc.go (hunk 1 above)
Comment (yobloxroute) [reply bxserum/provider/grpc.go]: 1. will think how to properly generalise it
2. basically we can use duration 0 for that, but flag will work too

File: bxserum/provider/grpc.go (hunk 2 above)
Comment (yobloxroute) [reply bxserum/provider/grpc.go:26]: I mean time, for hash, which contains mutex it's necessary

**Pull request closed without merging**